storage: TestSplitSnapshotRace_SnapshotWins is disabled

The following test appears to have failed:

[#21119](https://circleci.com/gh/cockroachdb/cockroach/21119):

```
I160809 16:10:59.496234 storage/engine/rocksdb.go:398  closing in-memory rocksdb instance
I160809 16:10:59.496523 storage/engine/rocksdb.go:398  closing in-memory rocksdb instance
I160809 16:10:59.499416 storage/engine/rocksdb.go:398  closing in-memory rocksdb instance
I160809 16:10:59.499679 storage/engine/rocksdb.go:398  closing in-memory rocksdb instance
I160809 16:10:59.500355 storage/engine/rocksdb.go:398  closing in-memory rocksdb instance
--- FAIL: TestSplitSnapshotRace_SnapshotWins (16.98s)
    <autogenerated>:12: storage/client_split_test.go:1050, condition failed to evaluate within 15s: expected [0 0 0 2 25 25], got [0 0 0 25 25 25]
=== RUN   TestStoreSplitTimestampCacheReadRace
I160809 16:10:59.511950 storage/engine/rocksdb.go:353  opening in memory rocksdb instance
W160809 16:10:59.514467 gossip/gossip.go:1015  not connected to cluster; use --join to specify a connected node
I160809 16:10:59.524527 storage/store.go:2732  system config not yet available
I160809 16:10:59.527705 storage/replica.go:2660  gossiping cluster id {"00000000-0000-0000-0000-000000000000"} from store 1, range 1
I160809 16:10:59.532568 storage/replica.go:2047  store=1:1 range=1 [/Min-/Max): applying command with forced error: storage/replica.go:1972: no-op on empty Raft entry
E160809 16:10:59.533413 storage/replica.go:2082  store=1:1 range=1 [/Min-/Max): error executing raft command: storage/replica.go:1972: no-op on empty Raft entry
I160809 16:10:59.534131 storage/replica_trigger.go:308  store=1:1 range=1 [/Min-/Max): new range lease replica {1 1 1} 1970-01-01 00:00:00 +0000 UTC 900.000123ms following replica {0 0 0} 1970-01-01 00:00:00 +0000 UTC 0s [physicalTime=1970-01-01 00:00:00.000000123 +0000 UTC]
I160809 16:10:59.534206 storage/replica.go:2682  gossiping sentinel from store 1, range 1
--
I160809 16:11:08.725458 stopper.go:408  quiesceing; tasks left:
18     gossip/infostore.go:290
I160809 16:11:08.725584 stopper.go:408  quiesceing; tasks left:
17     gossip/infostore.go:290
--- PASS: Example_rebalancing (0.95s)
FAIL
FAIL    github.com/cockroachdb/cockroach/storage    86.524s
=== RUN   TestBatchBasics
I160809 16:11:18.274805 storage/engine/rocksdb.go:353  opening in memory rocksdb instance
I160809 16:11:18.277250 storage/engine/rocksdb.go:398  closing in-memory rocksdb instance
--- PASS: TestBatchBasics (0.01s)
=== RUN   TestBatchRepr
I160809 16:11:18.292548 storage/engine/rocksdb.go:353  opening in memory rocksdb instance
I160809 16:11:18.294597 storage/engine/rocksdb.go:398  closing in-memory rocksdb instance
--- PASS: TestBatchRepr (0.01s)
=== RUN   TestBatchGet
I160809 16:11:18.302996 storage/engine/rocksdb.go:353  opening in memory rocksdb instance

```

Please assign, take a look and update the issue accordingly.

--- a/pkg/storage/client_split_test.go
+++ b/pkg/storage/client_split_test.go
@@ -1029,7 +1029,6 @@ func TestSplitSnapshotRace_SplitWins(t *testing.T) {
 // so it still has a conflicting range.
 func TestSplitSnapshotRace_SnapshotWins(t *testing.T) {
 	defer leaktest.AfterTest(t)()
-	t.Skip("#8416")
 	runSetupSplitSnapshotRace(t, func(mtc *multiTestContext, leftKey, rightKey roachpb.Key) {
 		// Bring the right range up first.
 		for i := 3; i <= 5; i++ {

--- a/pkg/storage/replica_command.go
+++ b/pkg/storage/replica_command.go
@@ -2994,22 +2994,13 @@ func (r *Replica) changeReplicasTrigger(
 	ctx context.Context, batch engine.Batch, change *roachpb.ChangeReplicasTrigger,
 ) ProposalData {
 	var pd ProposalData
-	// If we're removing the current replica, add it to the range GC queue.
-	if change.ChangeType == roachpb.REMOVE_REPLICA && r.store.StoreID() == change.Replica.StoreID {
-		// This wants to run as late as possible, maximizing the chances
-		// that the other nodes have finished this command as well (since
-		// processing the removal from the queue looks up the Range at the
-		// lease holder, being too early here turns this into a no-op).
-		pd.addToReplicaGCQueue = true
-	} else {
-		// After a successful replica addition or removal check to see if the
-		// range needs to be split. Splitting usually takes precedence over
-		// replication via configuration of the split and replicate queues, but
-		// if the split occurs concurrently with the replicas change the split
-		// can fail and won't retry until the next scanner cycle. Re-queuing
-		// the replica here removes that latency.
-		pd.maybeAddToSplitQueue = true
-	}
+	// After a successful replica addition or removal check to see if the
+	// range needs to be split. Splitting usually takes precedence over
+	// replication via configuration of the split and replicate queues, but
+	// if the split occurs concurrently with the replicas change the split
+	// can fail and won't retry until the next scanner cycle. Re-queuing
+	// the replica here removes that latency.
+	pd.maybeAddToSplitQueue = true
 
 	// Gossip the first range whenever the range descriptor changes. We also
 	// gossip the first range whenever the lease holder changes, but that might

--- a/pkg/storage/replica_proposal.go
+++ b/pkg/storage/replica_proposal.go
@@ -82,9 +82,6 @@ type LocalProposalData struct {
 	// The downstream-of-Raft logic does not exist at time of writing.
 	leaseMetricsResult *bool
 
-	// TODO(tschottdorf): there is no need to ever have these actions below
-	// taken on the followers, correct?
-
 	// When set (in which case we better be the first range), call
 	// gossipFirstRange if the Replica holds the lease.
 	gossipFirstRange bool
@@ -92,8 +89,6 @@ type LocalProposalData struct {
 	maybeGossipSystemConfig bool
 	// Call maybeAddToSplitQueue.
 	maybeAddToSplitQueue bool
-	// Call maybeAddToReplicaGCQueue.
-	addToReplicaGCQueue bool
 	// Call maybeGossipNodeLiveness with the specified Span, if set.
 	maybeGossipNodeLiveness *roachpb.Span
 }
@@ -238,7 +233,6 @@ func (p *ProposalData) MergeAndDestroy(q ProposalData) error {
 	coalesceBool(&p.gossipFirstRange, &q.gossipFirstRange)
 	coalesceBool(&p.maybeGossipSystemConfig, &q.maybeGossipSystemConfig)
 	coalesceBool(&p.maybeAddToSplitQueue, &q.maybeAddToSplitQueue)
-	coalesceBool(&p.addToReplicaGCQueue, &q.addToReplicaGCQueue)
 
 	if (q != ProposalData{}) {
 		log.Fatalf(context.TODO(), "unhandled ProposalData: %s", pretty.Diff(q, ProposalData{}))
@@ -512,6 +506,20 @@ func (r *Replica) handleReplicatedProposalData(
 			)
 		}
 		rpd.State.Desc = nil
+	}
+
+	if change := rpd.ChangeReplicas; change != nil {
+		if change.ChangeType == roachpb.REMOVE_REPLICA &&
+			r.store.StoreID() == change.Replica.StoreID {
+			// This wants to run as late as possible, maximizing the chances
+			// that the other nodes have finished this command as well (since
+			// processing the removal from the queue looks up the Range at the
+			// lease holder, being too early here turns this into a no-op).
+			if _, err := r.store.replicaGCQueue.Add(r, replicaGCPriorityRemoved); err != nil {
+				// Log the error; the range should still be GC'd eventually.
+				log.Errorf(ctx, "unable to add to replica GC queue: %s", err)
+			}
+		}
 		rpd.ChangeReplicas = nil
 	}
 
@@ -629,14 +637,6 @@ func (r *Replica) handleLocalProposalData(
 			log.Infof(ctx, "unable to gossip first range: %s", err)
 		}
 		lpd.gossipFirstRange = false
-	}
-
-	if lpd.addToReplicaGCQueue {
-		if _, err := r.store.replicaGCQueue.Add(r, replicaGCPriorityRemoved); err != nil {
-			// Log the error; the range should still be GC'd eventually.
-			log.Errorf(ctx, "unable to add to replica GC queue: %s", err)
-		}
-		lpd.addToReplicaGCQueue = false
 	}
 
 	if lpd.maybeAddToSplitQueue {

